clients/auth.go

```diff
@@ -0,0 +1,257 @@
+		retVal := RestrictedTokens{}
+		if err := json.NewDecoder(res.Body).Decode(&retVal); err != nil {
+			log.Println(err)
+			return nil, &status.StatusError{
+				Status: status.NewStatusf(res.StatusCode, "ListUserRestrictedTokens Unable to parse response[%s]", req.URL),
+			}
+		}
+		return retVal, nil
+	} else if res.StatusCode == 404 {
+		return nil, nil
+	} else {
+		return nil, &status.StatusError{
+			Status: status.NewStatusf(res.StatusCode, "Unexpected response code from service[%s]", req.URL),
+		}
+	}
+}
+
+// CreateRestrictedToken creates a restricted token for a given user
+func (client *AuthClient) CreateRestrictedToken(userID string, expirationTime time.Time, paths []string, token string) (*RestrictedToken, error) {
+	host := client.getHost()
+	if host == nil {
+		return nil, errors.New("No known auth hosts")
+	}
+	host.Path = path.Join(host.Path, "v1", "users", userID, "restricted_tokens")
+
+	payload := RestrictedTokenCreate{
+		Paths:          &paths,
+		ExpirationTime: &expirationTime,
+	}
+
+	if jsonToken, err := json.Marshal(payload); err != nil {
+		return nil, fmt.Errorf("unable to marshal payload: %w", err)
+	} else {
+		req, _ := http.NewRequest("POST", host.String(), bytes.NewBuffer(jsonToken))
+		req.Header.Add("x-tidepool-session-token", token)
+
+		res, err := client.httpClient.Do(req)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't create restricted token")
+		}
+		defer res.Body.Close()
+
+		switch res.StatusCode {
+		case http.StatusCreated:
+			var td RestrictedToken
+			if err = json.NewDecoder(res.Body).Decode(&td); err != nil {
+				log.Println("Error parsing JSON results", err)
+				return nil, nil
+			}
+			return &td, nil
+		default:
+			return nil, &status.StatusError{
+				Status: status.NewStatusf(res.StatusCode, "Unexpected response code from service[%s]", req.URL),
+			}
+		}
+	}
+}
+
+// UpdateRestrictedToken updates a restricted token
+func (client *AuthClient) UpdateRestrictedToken(tokenID string, expirationTime time.Time, paths []string, token string) (*RestrictedToken, error) {
+	host := client.getHost()
+	if host == nil {
+		return nil, errors.New("No known auth hosts")
+	}
+	host.Path = path.Join(host.Path, "v1", "restricted_tokens", tokenID)
+
+	payload := RestrictedTokenUpdate{
+		Paths:          &paths,
+		ExpirationTime: &expirationTime,
+	}
+
+	if jsonToken, err := json.Marshal(payload); err != nil {
+		return nil, fmt.Errorf("unable to marshal payload: %w", err)
+	} else {
+		req, _ := http.NewRequest("PUT", host.String(), bytes.NewBuffer(jsonToken))
+		req.Header.Add("x-tidepool-session-token", token)
+
+		res, err := client.httpClient.Do(req)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't update restricted token")
+		}
+		defer res.Body.Close()
+
+		switch res.StatusCode {
+		case http.StatusOK:
+			var td RestrictedToken
+			if err = json.NewDecoder(res.Body).Decode(&td); err != nil {
+				log.Println("Error parsing JSON results", err)
+				return nil, nil
```

Comment:
I think you should return the error here instead of just logging and ignoring it.

Reply:
Done